src/java_bytecode/java_types.h

```diff
@@ -30,6 +31,30 @@ class java_class_typet:public class_typet
   {
     return set(ID_access, access);
   }
+
+  typedef std::vector<symbol_exprt> java_lambda_method_handlest;
+
+  const java_lambda_method_handlest &lambda_method_handles() const
+  {
+    return (const java_lambda_method_handlest &)find(
+             ID_java_lambda_method_handles)
+      .get_sub();
+  }
+
+  java_lambda_method_handlest &lambda_method_handles()
+  {
+    return (java_lambda_method_handlest &)add(ID_java_lambda_method_handles)
+      .get_sub();
+  }
+
+  void add_lambda_method_handle(const irep_idt &identifier)
+  {
+    lambda_method_handles().emplace_back(identifier);
```

Comment:
This doesn't appear quite right to be, due to the mix an match of move semantics and const references. My understanding is that `emplace_back` will move its parameter into the vector, leaving the original copy it was passed in, in an invalid/undefined state. However the type of `identifier` is const reference, which means we are promising not to update it.

Reply:
`emplace_back` takes arguments to pass to the constructor of the vector type (in this case `const irep_idt &identifier` as an input to `symbol_exprt` constructor):
http://en.cppreference.com/w/cpp/container/vector/emplace_back
I don't think the identifier is left in an invalid/undefined state, maybe the original copy of `symbol_exprt` that is moved to the vector? Although the above source says that the new element is constructed 'in-place at the location provided by the container'

Reply:
Ok, thank you for explaining how this works, I was wondering where the conversion from irep_idt to symbol_exprt was happening. I had previously been reading about `&&` as part of move semantics. Why is the type of the `identifier` parameter not `irep_idt &&identifier` to match the type needed by `emplace_back`?

Reply:
`emplace_back` and `std::move` are different things, with similar aims. `emplace_back` just tries to avoid constructing a temporary object to pass as an argument, copy-constructing an object to go in the collection, and then destructing the original temporary. It can only be used to create a new object, but it does it directly in the place it's needed.